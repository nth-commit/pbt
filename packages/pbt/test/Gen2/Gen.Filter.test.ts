import fc from 'fast-check';
import * as dev from './srcShim';
import * as domainGen from './Helpers/domainGen';

test('sample(gen.filter(false)) *throws* exhausted', () => {
  fc.assert(
    fc.property(domainGen.sampleConfig(), domainGen.gen(), (config, gen) => {
      const genFiltered = gen.filter(() => false);

      expect(() => dev.sample(genFiltered, config)).toThrow('Exhausted after');
    }),
  );
});

test('sample(gen.filter(size > 50)) *produces* values *because* it resizes itself after a failed predicate', () => {
  fc.assert(
    fc.property(domainGen.sampleConfig(), (config) => {
      const gen = dev.Gen.create(
        (_, size) => size,
        dev.Shrink.none(),
        (size) => size,
      );
      const genFiltered = gen.filter((size) => size > 50);

      expect(() => dev.sample(genFiltered, config)).not.toThrow();
    }),
  );
});

test('sample(gen.map(f)) *produces* values, x, where f(x) = true', () => {
  fc.assert(
    fc.property(domainGen.sampleConfig(), domainGen.gen(), domainGen.predicate(), (config, gen, f) => {
      const genFiltered = gen.filter(f);

      const sample = dev.sampleTrees(genFiltered, config);

      for (const tree of sample.values) {
        expect(f(tree.node.value)).toEqual(true);
        for (const shrink of dev.GenTree.traverseGreedy(tree)) {
          expect(f(shrink.value)).toEqual(true);
        }
      }
    }),
  );
});
