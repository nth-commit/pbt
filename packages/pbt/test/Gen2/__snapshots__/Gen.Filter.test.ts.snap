// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot: 0 1`] = `"0 (c = 0)"`;

exports[`snapshot: 1 1`] = `"0 (c = 0)"`;

exports[`snapshot: 2 1`] = `
"2 (c = 20)
..0 (c = 0)"
`;

exports[`snapshot: 3 1`] = `"0 (c = 0)"`;

exports[`snapshot: 4 1`] = `
"4 (c = 40)
..0 (c = 0)
..2 (c = 20)
....0 (c = 0)"
`;

exports[`snapshot: 5 1`] = `
"2 (c = 20)
..0 (c = 0)"
`;

exports[`snapshot: 6 1`] = `
"4 (c = 40)
..0 (c = 0)
..2 (c = 20)
....0 (c = 0)"
`;

exports[`snapshot: 7 1`] = `
"4 (c = 40)
..0 (c = 0)
..2 (c = 20)
....0 (c = 0)"
`;

exports[`snapshot: 8 1`] = `
"8 (c = 80)
..0 (c = 0)
..4 (c = 40)
....0 (c = 0)
....2 (c = 20)
......0 (c = 0)
..6 (c = 60)
....0 (c = 0)
....4 (c = 40)
......0 (c = 0)
......2 (c = 20)
........0 (c = 0)"
`;

exports[`snapshot: 9 1`] = `
"4 (c = 40)
..0 (c = 0)
..2 (c = 20)
....0 (c = 0)"
`;

exports[`snapshot: 10 1`] = `
"10 (c = 100)
..0 (c = 0)
..8 (c = 80)
....0 (c = 0)
....4 (c = 40)
......0 (c = 0)
......2 (c = 20)
........0 (c = 0)
....6 (c = 60)
......0 (c = 0)
......4 (c = 40)
........0 (c = 0)
........2 (c = 20)
..........0 (c = 0)"
`;
